public/js/src/module/form-cache.js

```diff
@@ -25,21 +44,80 @@ function init( survey ) {
         .then( _setResetListener );
 }
 
+/**
+ * @param {Survey} survey
+ * @return Survey
+ */
 function get( survey ) {
     return store.survey.get( survey.enketoId );
 }
 
+/**
+ * @param {Survey} survey
+ * @return {Promise<Survey>}
+ */
 function set( survey ) {
     return connection.getFormParts( survey )
         .then( _swapMediaSrc )
         .then( _addBinaryDefaultsAndUpdateModel )
         .then( store.survey.set );
 }
 
+/**
+ *
+ * @param { string } enketoId
+ * @return { Promise<EnketoRecord | undefined> }
+ */
+function getLastSavedRecord( enketoId ) {
+    return store.survey.get( enketoId )
+        .then( survey => {
+            if ( survey != null ) {
+                return survey.lastSavedRecord;
+            }
+        } );
+}
+
+/**
+ *
+ * @param { string } enketoId
+ * @param { EnketoRecord } lastSavedRecord
+ * @return { Promise<Survey> }
+ */
+function setLastSavedRecord( enketoId, lastSavedRecord ) {
+    return store.survey.get( enketoId )
+        .then( survey => {
+            if ( encryptor.isEncryptionEnabled( survey ) ) {
+                return Promise.resolve( survey );
+            }
+
+            if ( Array.isArray( survey.externalData ) ) {
+                const { xml } = lastSavedRecord;
+
+                survey.externalData.forEach( item => {
+                    if ( item.src === LAST_SAVED_VIRTUAL_ENDPOINT ) {
+                        item.xml = parser.parseFromString( xml, 'text/xml' );
+                    }
+                } );
+            }
+
+            const update = Object.assign( {}, survey, { lastSavedRecord } );
+
+            return store.survey.update( update );
+        } );
+}
+
+/**
+ * @param {Survey} survey
+ * @return {Promise<void>}
+ */
 function remove( survey ) {
     return store.survey.remove( survey.enketoId );
 }
 
+/**
+ * @param {Survey} survey
+ * @return {PRomise<Survey>}
```

Comment:
Spotted this typo too late 😬. I’m on mobile right now so I’ll fix this and anything else I spot in the morning.